### ISSUE TITLE
build fail on Linux

### ISSUE THREAD

**Issue by davephillips** [Rcomian/jsfx-vcv#1]
Fedora 23 Linux
Rack 0.6.2b (from source)
GCC 5.3.1

Greetings ! I love the idea for this module, but alas, my build attempt fails here:

make dep
...
/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp: In member function ‘void JsusFxPathLibrary_Basic::addSearchPath(const string&)’:
/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:351:12: error: ‘const string {aka const class std::__cxx11::basic_string<char>}’ has no member named ‘back’
  if ( path.back() == '/' || path.back() == '\\' )
            ^
/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:351:34: error: ‘const string {aka const class std::__cxx11::basic_string<char>}’ has no member named ‘back’
  if ( path.back() == '/' || path.back() == '\\' )
                                  ^
/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp: In static member function ‘static bool JsusFxPathLibrary_Basic::fileExists(const string&)’:
/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:358:27: error: no matching function for call to ‘std::basic_ifstream<char>::basic_ifstream(const string&)’
  std::ifstream is(filename);
                           ^
In file included from /home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:47:0:
/usr/include/c++/5.3.1/fstream:495:7: note: candidate: std::basic_ifstream<_CharT, _Traits>::basic_ifstream(const char*, std::ios_base::openmode) [with _CharT = char; _Traits = std::char_traits<char>; std::ios_base::openmode = std::_Ios_Openmode]
       basic_ifstream(const char* __s, ios_base::openmode __mode = ios_base::in)
       ^
/usr/include/c++/5.3.1/fstream:495:7: note:   no known conversion for argument 1 from ‘const string {aka const std::__cxx11::basic_string<char>}’ to ‘const char*’
In file included from /home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:47:0:
/usr/include/c++/5.3.1/fstream:481:7: note: candidate: std::basic_ifstream<_CharT, _Traits>::basic_ifstream() [with _CharT = char; _Traits = std::char_traits<char>]
       basic_ifstream() : __istream_type(), _M_filebuf()
       ^
/usr/include/c++/5.3.1/fstream:481:7: note:   candidate expects 0 arguments, 1 provided
In file included from /home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:47:0:
/usr/include/c++/5.3.1/fstream:455:11: note: candidate: std::basic_ifstream<char>::basic_ifstream(const std::basic_ifstream<char>&)
     class basic_ifstream : public basic_istream<_CharT, _Traits>
           ^
/usr/include/c++/5.3.1/fstream:455:11: note:   no known conversion for argument 1 from ‘const string {aka const std::__cxx11::basic_string<char>}’ to ‘const std::basic_ifstream<char>&’
/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp: In member function ‘virtual bool JsusFxPathLibrary_Basic::resolveImportPath(const string&, const string&, std::__cxx11::string&)’:
/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:372:35: warning: range-based ‘for’ loops only available with -std=c++11 or -std=gnu++11
  for ( std::string & searchPath : searchPaths ) {
                                   ^
/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp: In member function ‘virtual std::istream* JsusFxPathLibrary_Basic::open(const string&)’:
/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:391:48: error: no matching function for call to ‘std::basic_ifstream<char>::basic_ifstream(const string&)’
  std::ifstream *stream = new std::ifstream(path);
                                                ^
In file included from /home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:47:0:
/usr/include/c++/5.3.1/fstream:495:7: note: candidate: std::basic_ifstream<_CharT, _Traits>::basic_ifstream(const char*, std::ios_base::openmode) [with _CharT = char; _Traits = std::char_traits<char>; std::ios_base::openmode = std::_Ios_Openmode]
       basic_ifstream(const char* __s, ios_base::openmode __mode = ios_base::in)
       ^
/usr/include/c++/5.3.1/fstream:495:7: note:   no known conversion for argument 1 from ‘const string {aka const std::__cxx11::basic_string<char>}’ to ‘const char*’
In file included from /home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:47:0:
/usr/include/c++/5.3.1/fstream:481:7: note: candidate: std::basic_ifstream<_CharT, _Traits>::basic_ifstream() [with _CharT = char; _Traits = std::char_traits<char>]
       basic_ifstream() : __istream_type(), _M_filebuf()
       ^
/usr/include/c++/5.3.1/fstream:481:7: note:   candidate expects 0 arguments, 1 provided
In file included from /home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b/src/jsusfx.cpp:47:0:
/usr/include/c++/5.3.1/fstream:455:11: note: candidate: std::basic_ifstream<char>::basic_ifstream(const std::basic_ifstream<char>&)
     class basic_ifstream : public basic_istream<_CharT, _Traits>
           ^
/usr/include/c++/5.3.1/fstream:455:11: note:   no known conversion for argument 1 from ‘const string {aka const std::__cxx11::basic_string<char>}’ to ‘const std::basic_ifstream<char>&’
CMakeFiles/jsusfx.dir/build.make:66: recipe for target 'CMakeFiles/jsusfx.dir/jsusfx.cpp.o' failed
make[3]: *** [CMakeFiles/jsusfx.dir/jsusfx.cpp.o] Error 1
make[3]: Leaving directory '/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b'
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/jsusfx.dir/all' failed
make[2]: *** [CMakeFiles/jsusfx.dir/all] Error 2
make[2]: Leaving directory '/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b'
Makefile:83: recipe for target 'all' failed
make[1]: *** [all] Error 2
make[1]: Leaving directory '/home/dlphilp/src/Rack-062/plugins/jsfx-vcv/dep/jsusfx-0.4.0b1b'
Makefile:30: recipe for target 'lib/jsusfx.a' failed
make: *** [lib/jsusfx.a] Error 2

It seems that it wants -std=cxx11 somewhere but I'm not sure where that will go in a cmake system. Any suggestions are vastly appreciated. :)

**Comment by Rcomian** [comment]
Yeah building the jsusfx code is pretty tricky, I've found.
It might be worth cloning that code directly and trying to get it to build: https://github.com/asb2m10/jsusfx, then if you can get that environment right, it should build within the module project.

I only cloned the repository to get access to slightly later features than were in their last release and to add the -fPIC flag to jsusfx/src/CMakeLists.txt. If you need to add extra flags that might be the place to do so. I'm guessing tho.

(I didn't write the actual engine, that would have been way too much work :P)
If it helps at all, I can tell you what versions of tools I'm running in case that matters:

* gcc 7.3.0
* make 4.2.1
* cmake 3.9.6
* ld 2.30.0
* php 7.1.18
* nasm 2.13.01

If you like I can make a release with linux binaries, to save you from the pain of building. I can't get it to build on windows at all, which is why I haven't made a release yet.

Also, how on earth did you find this project? I'm very glad you did, but I haven't told anyone about it :p

**Comment by Rcomian** [comment]
In fact you can try this here: https://www.dropbox.com/s/63816r3djvf5zbn/jsfx-0.6.0-lin.zip?dl=0

**Comment by davephillips** [comment]
Greetings ! Hey, thanks for the tips, I did eventually build and test the plugin. It's very cool, I'll test some of the fx tonight. You asked how I found your project: I'm a habitual git crawler, not a developer, but I know how to compile and debug from source code. Thank you for the binary, I haven't tested it yet. Btw, I thought you might like to see this:
![jsfx-vcv-linux](https://user-images.githubusercontent.com/7815938/45589879-9ae80880-b8fb-11e8-8f0d-94d326ad05da.png)
